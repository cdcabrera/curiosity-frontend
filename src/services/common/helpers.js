import _camelCase from 'lodash/camelCase';
import _isPlainObject from 'lodash/isPlainObject';

const camelCase = obj => {
  if (Array.isArray(obj)) {
    return obj.map(camelCase);
  }

  if (_isPlainObject(obj)) {
    const updatedObj = {};

    Object.entries(obj).forEach(([key, val]) => {
      updatedObj[_camelCase(key)] = camelCase(val);
    });

    return updatedObj;
  }

  return obj;
};

const schemaResponse = ({ convert = true, id = null, response, schema }) => {
  const { value, error = { details: [] } } = schema.validate(response, { convert });

  if (error.details.length) {
    throw new Error(
      `Schema validation for ${id || '...'}: ${error.details
        .map(({ message, type }) => `${message}:${type}`)
        .join(', ')}`
    );
  }

  return camelCase(value);
};

const serviceHelpers = {
  camelCase,
  schemaResponse
};

export { serviceHelpers as default, camelCase, schemaResponse };
