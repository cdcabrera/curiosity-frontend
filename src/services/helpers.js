import _camelCase from 'lodash/camelCase';

const schemaResponse = (response, schema, id = null) => {
  const { value, error = { details: [] } } = schema.validate(response);

  if (error.details.length) {
    throw new Error(
      `Schema validation for ${id}: ${error.details.map(({ message, type }) => `${message}:${type}`).join(', ')}`
    );
  }

  const updatedValue = {};
  Object.keys(value).forEach(key => {
    updatedValue[_camelCase(key).trim()] = value[key];
  });

  return updatedValue;
};

const serviceHelpers = {
  schemaResponse
};

export { serviceHelpers as default, schemaResponse };
