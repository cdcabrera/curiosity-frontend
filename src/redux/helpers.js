import _camelCase from 'lodash/camelCase';
import _snakeCase from 'lodash/snakeCase';
import { helpers } from '../common';

/**
 * @memberof Helpers
 * @module ReduxHelpers
 */

/**
 * Set an API query based on specific API "acceptable values" schema.
 *
 * @param {object} values
 * @param {object} schema
 * @param {*} [initialValue]
 * @returns {object}
 */
const setApiQuery = (values, schema, initialValue) => {
  const generated = {};
  const schemaArr = (schema && Object.values(schema)) || [];

  schemaArr.forEach(value => {
    if (initialValue === undefined) {
      if (value in values) {
        generated[value] = values?.[value];
      }
    } else {
      generated[value] = values?.[value] || initialValue;
    }
  });

  return generated;
};

// ToDo: research applying a maintained schema map/normalizer such as, normalizr
/**
 * Apply a set of schemas using either an array of objects in the
 * form of [{ madeUpKey: 'some_api_key' }], or an array of arrays
 * in the form of [['some_api_key','another_api_key']]
 *
 * @param {Array} schemas
 * @param {*} [initialValue]
 * @returns {Array}
 */
const setResponseSchemas = (schemas = [], initialValue) =>
  schemas.map(schema => {
    const generated = {};
    const arr = (Array.isArray(schema) && schema) || Object.values(schema);

    arr.forEach(value => {
      generated[value] = initialValue;
    });

    return generated;
  });

/**
 * Normalize an API response.
 *
 * @param {*} responses
 * @param {object} responses.response
 * @param {object} responses.response.schema
 * @param {Array|object} responses.response.data
 * @param {string} responses.response.keyCase
 * @param {Function} responses.response.customResponseEntry
 * @param {Function} responses.response.customResponseValue
 * @param {string} responses.response.keyPrefix
 * @returns {Array}
 */
const setNormalizedResponse = (...responses) => {
  const parsedResponses = [];

  responses.forEach(
    ({ schema = {}, data, customResponseEntry, customResponseValue, keyPrefix: prefix, keyCase = 'camel' }) => {
      const isArray = Array.isArray(data);
      const updatedData = (isArray && data) || (data && [data]) || [];
      const [generatedSchema = {}] = setResponseSchemas([schema]);
      const parsedResponse = [];

      updatedData.forEach((value, index) => {
        const generateReflectedData = ({
          dataObj,
          keyPrefix = '',
          keyCaseType,
          customEntry,
          customValue = null,
          update = helpers.noop
        }) => {
          let updatedDataObj = {};

          Object.entries(dataObj).forEach(([dataObjKey, dataObjValue]) => {
            let casedDataObjKey;

            switch (keyCaseType) {
              case 'camel':
                casedDataObjKey = _camelCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'snake':
                casedDataObjKey = _snakeCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'default':
              default:
                casedDataObjKey = `${dataObjKey}`.trim();
                break;
            }

            let val = dataObjValue;

            if (typeof val === 'number') {
              val = (Number.isInteger(val) && Number.parseInt(val, 10)) || Number.parseFloat(val) || val;
            }

            if (typeof customValue === 'function') {
              updatedDataObj[casedDataObjKey] = customValue({ data: dataObj, key: dataObjKey, value: val, index });
            } else {
              updatedDataObj[casedDataObjKey] = val;
            }
          });

          if (typeof customEntry === 'function') {
            updatedDataObj = customEntry(updatedDataObj, index);
          }

          update(updatedDataObj);
        };

        generateReflectedData({
          keyPrefix: prefix,
          dataObj: { ...generatedSchema, ...value },
          keyCaseType: keyCase,
          customEntry: customResponseEntry,
          customValue: customResponseValue,
          update: generatedData => parsedResponse.push(generatedData)
        });
      });

      parsedResponses.push(parsedResponse);
    }
  );

  return parsedResponses;
};

const reduxHelpers = {
  setApiQuery,
  setResponseSchemas,
  setNormalizedResponse
};

export { reduxHelpers as default, reduxHelpers };
