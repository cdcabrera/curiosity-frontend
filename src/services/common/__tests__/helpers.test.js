import { serviceHelpers } from '../helpers';

describe('Service Helpers', () => {
  it('should have specific functions', () => {
    expect(serviceHelpers).toMatchSnapshot('serviceHelpers');
  });

  it('should support camel casing obj keys', () => {
    expect(
      serviceHelpers.camelCase({
        lorem_ipsum: [
          {
            dolor_sit: [
              {
                hello_world: 1
              }
            ]
          }
        ]
      })
    ).toMatchSnapshot('camelCase');
  });

  it('should support applying a normalize function for schemas', () => {
    const mockSchema = jest.fn();

    serviceHelpers.responseNormalize({ hello: 'world' }, mockSchema);

    expect(mockSchema.mock.calls).toMatchSnapshot('responseNormalize, success');

    mockSchema.mockClear();
  });

  it('should attempt to apply a Joi schema to a response', () => {
    const mockValidate = jest.fn().mockImplementation(response => ({ value: response }));

    serviceHelpers.schemaResponse({ schema: { validate: mockValidate }, response: { lorem_ipsum: 'dolor_sit' } });
    expect(mockValidate.mock.calls).toMatchSnapshot('schemaResponse, parameters');

    expect(
      serviceHelpers.schemaResponse({
        schema: { validate: mockValidate },
        response: { lorem_ipsum: 'dolor_sit' },
        casing: 'camel'
      })
    ).toMatchSnapshot('schemaResponse, camelCasing');
  });
});
