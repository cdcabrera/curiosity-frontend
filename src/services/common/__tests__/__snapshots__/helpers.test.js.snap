// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Helpers should attempt to apply a Joi schema to a response: schemaResponse, camelCasing 1`] = `
Object {
  "loremIpsum": "dolor_sit",
}
`;

exports[`Service Helpers should attempt to apply a Joi schema to a response: schemaResponse, parameters 1`] = `
Array [
  Array [
    Object {
      "lorem_ipsum": "dolor_sit",
    },
    Object {
      "convert": true,
    },
  ],
]
`;

exports[`Service Helpers should have specific functions: serviceHelpers 1`] = `
Object {
  "camelCase": [Function],
  "responseNormalize": [Function],
  "schemaResponse": [Function],
}
`;

exports[`Service Helpers should support applying a normalize function for schemas: responseNormalize, success 1`] = `
Array [
  Array [
    Object {
      "hello": "world",
    },
  ],
]
`;

exports[`Service Helpers should support camel casing obj keys: camelCase 1`] = `
Object {
  "loremIpsum": Array [
    Object {
      "dolorSit": Array [
        Object {
          "helloWorld": 1,
        },
      ],
    },
  ],
}
`;
